docs/examples/starlette/simple_starlette_app.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+import os
+
+from typing import AsyncGenerator
+
+from starlette.applications import Starlette
+from starlette.middleware import Middleware
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+from starlette.routing import Route
+
+import svcs
+
+
+config = {"db_url": os.environ.get("DB_URL", "sqlite:///:memory:")}
+
+
+class Database:
+    @classmethod
+    async def connect(cls, db_url: str) -> Database:
+        ...
+        return Database()
+
+    async def get_user(self, user_id: int) -> dict[str, str]:
+        return {}  # not interesting here
+
+
+async def get_user(request: Request) -> JSONResponse:
+    db = await svcs.starlette.aget(request, Database)
+
+    try:
+        return JSONResponse(
+            {"data": await db.get_user(request.path_params["user_id"])}
+        )
+    except Exception as e:
+        return JSONResponse({"oh no": e.args[0]})
+
+
+@svcs.starlette.lifespan
+async def lifespan(
+    app: Starlette, registry: svcs.Registry
+) -> AsyncGenerator[dict[str, object], None]:
+    async def connect_to_db() -> Database:
+        return await Database.connect(config["db_url"])
+
+    registry.register_factory(Database, connect_to_db)
+
+    yield {"your": "other stuff"}
```

Comment:
Without `svcs` I would do:

```py
async def get_user(request: Request) -> JSONResponse:
    db = request.state.database

    try:
        return JSONResponse(
            {"data": await db.get_user(request.path_params["user_id"])}
        )
    except Exception as e:
        return JSONResponse({"oh no": e.args[0]})


@svcs.starlette.lifespan
async def lifespan(
    app: Starlette, registry: svcs.Registry
) -> AsyncGenerator[dict[str, object], None]:
    database = await Database.connect(config["db_url"])

    yield {"your": "other stuff", "db": database}
```

What's the benefit of using `svcs` here?

Reply:
that's a wider question I'm trying to answer in the docs, but I probably should try to centralize more.

1. your database is app-scoped, no?
2. your database is always connected; svcs allows you to obtain dependencies only when you need them (e.g. `if form.valid(): svcs_from(request).get(Database).do_db_stuff_with_form(form)`. You either have to pass in a pool that you manage manually or create DB connections that you won't need. The svcs way is both convenient and flexible.
3. your solution has no concept of lifecycles; *svcs* has lifecycle management for both the pool (on_close_registry) and the service (automatic close after request)
4. health checks!
5. idiomatic overwriting of services in tests including all the upsides from above; I find it at least awkward to have proper cleanups when monkeypatching around
6. Also, as I write on the why page: it gets a lot more interesting once you have more than one dependency to manage. Knowing where they live and accessing them in a consistent way has proven to make things easier – and easier to understand.

---

TBH I fully expect 3 people getting **really** excited about svcs because they've hit pain points 1–6 and the rest shrugging because they haven't.

Reply:
Oh and: 7. with Starlette one upside is also type-safety. `request.state.db` is Any. `aget(request, Database)` is Database to Mypy et al.

Reply:
Thanks @hynek :)